tests/pnr/test_hpwl.py

```diff
@@ -94,3 +119,359 @@ def test_gen_netlist_flip():
+        {
+          "abstract_template_name": "row",
+          "concrete_template_name": "row_0",
+          "fa_map": [
+            {
+              "actual": "x3",
+              "formal": "inp"
+            },
+            {
+              "actual": "out",
+              "formal": "out"
+            }
+          ],
+          "instance_name": "u3",
+          "transformation": {
+            "oX": 0,
+            "oY": 0,
+            "sX": 1,
+            "sY": 1
+          }
+        }
+      ],
+      "parameters": [
+        "inp",
+        "out"
+      ]
+    },
+    {
+      "abstract_name": "row",
+      "bbox": [
+        0,
+        0,
+        2480,
+        840
+      ],
+      "concrete_name": "row_0",
+      "constraints": [
+        {
+          "abut": false,
+          "constraint": "order",
+          "direction": "left_to_right",
+          "instances": [
+            "u0",
+            "u1",
+            "u2"
+          ]
+        },
+        {
+          "constraint": "same_template",
+          "instances": [
+            "u0",
+            "u1",
+            "u2"
+          ]
+        }
+      ],
+      "instances": [
+        {
+          "abstract_template_name": "slice",
+          "concrete_template_name": "slice_a",
+          "fa_map": [
+            {
+              "actual": "inp",
+              "formal": "inp"
+            },
+            {
+              "actual": "x1",
+              "formal": "out"
+            }
+          ],
+          "instance_name": "u0",
+          "transformation": {
+            "oX": 0,
+            "oY": 0,
+            "sX": 1,
+            "sY": 1
+          }
+        },
+        {
+          "abstract_template_name": "slice",
+          "concrete_template_name": "slice_a",
+          "fa_map": [
+            {
+              "actual": "x1",
+              "formal": "inp"
+            },
+            {
+              "actual": "x2",
+              "formal": "out"
+            }
+          ],
+          "instance_name": "u1",
+          "transformation": {
+            "oX": 880,
+            "oY": 0,
+            "sX": 1,
+            "sY": 1
+          }
+        },
+        {
+          "abstract_template_name": "slice",
+          "concrete_template_name": "slice_a",
+          "fa_map": [
+            {
+              "actual": "x2",
+              "formal": "inp"
+            },
+            {
+              "actual": "out",
+              "formal": "out"
+            }
+          ],
+          "instance_name": "u2",
+          "transformation": {
+            "oX": 1680,
+            "oY": 0,
+            "sX": 1,
+            "sY": 1
+          }
+        }
+      ],
+      "parameters": [
+        "inp",
+        "out"
+      ]
+    }
+  ]
+}
+"""
+
+    placement_verilog_d = json.loads(txt)
+
+
+    cn = 'matrix_0'
+    nets_d = gen_netlist( placement_verilog_d, cn)
+
+    assert 27584 == calculate_HPWL_from_placement_verilog_d( placement_verilog_d, cn, nets_d)
```

Comment:
I do some local transformations to reduce area by 9% and hpwl (calculated in a better way) by > 40%.
We need to improve the placer so it can achieve this placement. Requires probably:
- [ ] Using this HPWL calculation in the placer code
- [ ] Allowing flips on both the X and Y axes
- [ ] The SameTemplate constraint should also help reduce the search space